### DISCUSSION TITLE
Issue resolved on BQC problem where fails on timeout 'waiting for data'

### DISCUSSION THREAD

**Discussion by Sean-Steefel** [MicrosoftPremier/VstsExtensions#196] [General]
Hi @ReneSchumacher,

I wanted to reach out to confirm if this was a known issue.  Probably really a problem for the developer consuming this extension but felt like it was worth bubbling up.

Recently we encountered a repeating problem where the BQC would timeout on almost all of our build pipelines and therefor our builds were not reporting failures.  In analysis I found that our unit tests had a cyclical dependency and was actually running tests repeatedly in only the ADO build environment (VS 2022 - interactive did not present the problem)...  Tests would pass without issue -- however the BQC would then loop on 'waiting for data...' and then failout.

When I had the developers remove that cyclical dependency (basically a utility class within a test project of other tests) by elevating the utility class to it's own project -- thereby stopping the repetitive tests also fixed the BQC timeout!

I am hoping with this info others might find their own issues -- or perhaps BQC might be able to handle with a warning of cyclical dependencies?

**Comment by ReneSchumacher** [comment]
Hi @Sean-Steefel.

thanks for pointing that out. I'll try to build a repro and see if we can handle the issue. However, my guess is that the underlying issue is somewhere in Azure DevOps. BQC merely waits for all previous tasks to report that they're done, which is a status in the build timeline. I'm wondering how a task can finish (i.e., the process should have ended) but still be listed as running.

What we could do is skip waiting for tasks after a certain timeout. This would pose the risk of trying to read result data (logs or coverage data) before it has been uploaded to Azure DevOps, though.
